data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/authenticator/DynamicSaslClientCallbackHandler.java

```diff
@@ -0,0 +1,38 @@
+package org.opensearch.dataprepper.plugins.kafka.authenticator;
+
+import org.apache.kafka.common.security.authenticator.SaslClientCallbackHandler;
+
+import javax.security.auth.callback.Callback;
+import javax.security.auth.callback.NameCallback;
+import javax.security.auth.callback.PasswordCallback;
+import javax.security.auth.callback.UnsupportedCallbackException;
+
+public class DynamicSaslClientCallbackHandler extends SaslClientCallbackHandler {
+    private final DynamicBasicCredentialsProvider dynamicBasicCredentialsProvider = DynamicBasicCredentialsProvider
+            .getInstance();
+
+    @Override
+    public void handle(Callback[] callbacks) throws UnsupportedCallbackException {
+        super.handle(callbacks);
+        final BasicCredentials basicCredentials = dynamicBasicCredentialsProvider.getBasicCredentials();
+        for (final Callback callback: callbacks) {
+            if (callback instanceof NameCallback) {
+                NameCallback nc = (NameCallback) callback;
+                if (basicCredentials.getUsername() != null) {
+                    nc.setName(basicCredentials.getUsername());
+                } else {
+                    nc.setName(nc.getDefaultName());
+                }
+            } else if (callback instanceof PasswordCallback) {
+                PasswordCallback nc = (PasswordCallback) callback;
+                if (basicCredentials.getPassword() == null) {
+                    String errorMessage = "Could not login: the client is being asked for a password.";
+                    throw new UnsupportedCallbackException(callback, errorMessage);
+                }
+
+                final char[] password = basicCredentials.getPassword().toCharArray();
+                nc.setPassword(password);
+            }
```

Comment:
Are there more conditions here? Perhaps add a log line (WARN) for an else? If we do get another `Callback` subclass we might have very subtle errors.

Reply:
So all Callbacks are already handled in super.handle(callbacks). Here it is just an overwrite on those two callbacks

Reply:
I'm not clear on who is overriding this for only these two callbacks?

I'd still tend to think that adding a few lines to log a warning would be helpful in case either of these: 1) We add a bug that adds a new callback elsewhere, but don't support here; or 2) the Kafka implementation changes.

```
else {
  log.warn("Received an unexpected Callback type {}", callback.getClass());
}
```

Reply:
By inspecting their source code: https://github.com/a0x8o/kafka/blob/master/clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientCallbackHandler.java#L75-L98

I saw UnsupportedCallbackException would be thrown.